### PR TITLE
wip: v2

### PR DESCRIPTION
**Pull request by samparsky**

### CODE REVIEW

File: contracts/v2/Outpace.sol [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        
+        for(uint i = 0; i < bulkWithdraw.channels.length; i++) {
+            ChannelLibraryV2.Channel calldata channel = bulkWithdraw.channels[i];
+            bytes32 stateRoot = bulkWithdraw.stateRoots[i];
+            uint amountInTree = bulkWithdraw.amountInTrees[i];
+            bytes32[3][] calldata signature = bulkWithdraw.signatures[i];
+
+            bytes32 channelId = ChannelLibraryV2.hash(channel);
+            require(states[channelId] == ChannelLibraryV2.State.Active, "INVALID_STATE");
+            require(now <= channel.validUntil, "EXPIRED");
```
Comment (Ivshti) [inline contracts/v2/Outpace.sol:93]: we may be getting rid of that check: this is something that's described in the AIP

File: contracts/v2/Outpace.sol [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        
+            bytes32 hashToSign = keccak256(abi.encode(channelId, stateRoot));
+            require(ChannelLibraryV2.isSignedBySupermajority(channel, hashToSign, signature), "NOT_SIGNED_BY_VALIDATORS");
+        
+            bytes32 balanceLeaf = keccak256(abi.encode(msg.sender, amountInTree));
+            require(MerkleProof.isContained(balanceLeaf, bulkWithdraw.proofs[i], stateRoot), "BALANCELEAF_NOT_FOUND");
+		    
+            uint256 amountToWithdraw;
+            if (bulkWithdraw.amountWithdrawnPerChannel.length > 0 && (bulkWithdraw.amountWithdrawnPerChannel.length - 1) > i) {
```
Comment (Ivshti) [inline contracts/v2/Outpace.sol:102]: will explain this one in discord - matching by index won't work here

**Review by Ivshti — CHANGES_REQUESTED** [review]

**Comment by Ivshti** [comment]
Closing in favor of https://github.com/AdExNetwork/aips/issues/61

**Comment by samparsky** [comment]
ok, I will take a deeper look at it

**Pull request closed without merging**